# Parte 1: listas e _list comprehensions_

## Exercício 1

Escreva uma função `f_vetor(f,v)` que recebe uma função `f` e um vetor `v` (numa lista)
e retorna o vetor com `f` calculada em cada coordenada de `v`.
Em fórmulas, se $v = (v_i)$, então `f_vetor(f,v)` = $ (f(v_i)) $.

In [1]:
def f_vetor(f,v):
    return [f(x) for x in v]

In [2]:
import math
def sq(x): return x**2

Teste que sua função funciona na caixa abaixo (você pode também testar *outras* funções e vetores, e criar outras caixas!)

In [3]:
f_vetor(sq, [1,2])

[1, 4]

In [4]:
assert(f_vetor(sq, [1,3,5,7,9]) == [1, 9, 25, 49, 81])
assert(f_vetor(math.sin, [0,1,2,3]) == [0.0, 0.8414709848078965, 0.9092974268256817, 0.1411200080598672])
assert(f_vetor(math.ceil, [1.72, -8.29, 10.45, 9.1, 1.0, -4.1391213771510751]) == [2, -8, 11, 10, 1, -4])

In [5]:
assert(f_vetor(sq, []) == [])
assert(f_vetor(math.sin, []) == [])

Agora, alguns testes aleatórios

In [6]:
from numpy.random import random # Mais poderoso do que random.random

In [7]:
v = random(size=(10))
assert(f_vetor(math.sqrt, f_vetor(sq, v)) == list(v))
v = random(size=(10))
assert(f_vetor(math.sin, f_vetor(math.asin, v)) == list(v))
v = random(size=(30))
assert(f_vetor(math.sqrt, f_vetor(sq, v)) == list(v))

Explique porque no segundo teste a ordem das funções foi `math.sin` e depois `math.asin`, e não na ordem contrária.

# <font color='red'>NÃO ESQUECER!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!</font>

## Exercício 2

Escreva uma função que, dados dois vetores $u$ e $v$ (de novo, numa lista), "complete" o vetor `u` para que ele fique do tamanho do vetor `v`.
Se a dimensão de $u$ for maior do que a de $v$, a função deve lançar a exceção `ValueError`.

In [8]:
def completar(u,v):
    diff = len(v)-len(u)
    if (diff<0):
        raise ValueError
    return u+[0]*(diff)

De novo, teste a sua função antes dos asserts (se você quiser)

In [9]:
u = [1,2,3]
v1 = [math.pi]*10
v2 = [1,1,1,1,1,1]
v3 = random(3)
assert(completar(u,v1) == [1,2,3,0,0,0,0,0,0,0])
assert(completar(u,v2) == [1,2,3,0,0,0])
assert(completar(u,v3) == [1,2,3])

A seguir, a função `verify_exc` verifica se, ao chamar uma outra função (`f`) ocorre a exceção `ex`.

In [10]:
def verify_exc(f, ex):
    try:
        f()
        return False
    except ex:
        return True

In [11]:
def completar_menos():
    return completar(u, [1,2])

assert(verify_exc(completar_menos, ValueError) == True)

In [12]:
def completar_a_zero():
    return completar(u, [])

assert( completar([], [1,2,3]) == [0,0,0] )
assert( completar([], []) == [])
assert( verify_exc(completar_a_zero, ValueError) == True )

## Exercício 3:

Adapte a função `f_vetor` para uma função `f_matriz`.

In [13]:
def f_matriz(f,m):
    return [f_vetor(f,v) for v in m]

In [14]:
x = math.sin(1)
y = math.sin(2)
z = math.sin(3)
assert( f_matriz(math.sin, [[]]) == [[]])
assert( f_matriz(math.sin, [[1,2],[3,0]]) == [[x,y],[z,0.0]])

In [15]:
def teste_vetor():
    return f_matriz(math.sin, [1,2,3])

assert( verify_exc(teste_vetor, TypeError) )

# Parte 2: Recorrências

Escreva uma função **recursiva** que retorne o $n$-ésimo número de "Tribonacci", definido pelas equações
$$ T_0 = 0 \qquad T_1 = 1 \qquad T_2 = 1$$
$$ F_{n+3} = F_{n+2} + F_{n+1} + F_n. $$

In [16]:
def tribonacci(n):
    if (n==0): return 0
    if (n==1): return 1
    if (n==2): return 1
    return tribonacci(n-1)+tribonacci(n-2)+tribonacci(n-3)

Teste que a sua função funciona!

In [17]:
tribonacci(20)

66012

Agora, os testes que valem ponto!

In [18]:
assert(tribonacci(5) == 7)
assert(tribonacci(10) == 149)
assert(tribonacci(20) == 66012)

É sabido que a solução para uma recorrência linear é dada por uma combinação de potências.
No nosso caso, $T_n = A a^n + B b^n + C c^n$ onde $a$, $b$ e $c$ são as raízes de $x^3 = x^2 + x + 1$.

Verifique (numericamente) que $T_{n+1} / T_n$ converge. Para qual valor? De quantos dígitos você tem certeza?

In [29]:
def t(eps):
    tribarray = [1,1,2] #[t(1), t(2), t(3)]
    index_new = -1
    index_prev = 0
    index_curr = 1
    delta = 1
    prev = 1 #comeca como t(1)/t(2) = 1/1
    curr = 1
    counter = 1
    zero = False
    while (delta > eps or zero == True):
        index_new = (index_new+1)%3
        index_prev = (index_prev+1)%3
        index_curr = (index_curr+1)%3
        curr = tribarray[index_curr]/tribarray[index_prev]
        delta = abs(curr-prev)
        if (delta == 0): zero = not zero
        if (delta != 0 and zero): zero = False
        prev = curr
        tribarray[index_new] = tribarray[index_new]+tribarray[index_prev]+tribarray[index_curr]
        counter = counter+1
    print ("Para um epsilon =",eps)
    print (" n  | delta( T(n+1)/T(n) )")
    print (counter, " |", delta)
    print ("converge para ", curr)
    print ("")

In [30]:
def t_force(eps, min_iter):
    tribarray = [1,1,2] #[t(1), t(2), t(3)]
    index_new = -1
    index_prev = 0
    index_curr = 1
    delta = 1
    prev = 1 #comeca como t(1)/t(2) = 1/1
    curr = 1
    counter = 1
    while (delta > eps or counter < min_iter):
        index_new = (index_new+1)%3
        index_prev = (index_prev+1)%3
        index_curr = (index_curr+1)%3
        curr = tribarray[index_curr]/tribarray[index_prev]
        delta = abs(curr-prev)
        prev = curr
        tribarray[index_new] = tribarray[index_new]+tribarray[index_prev]+tribarray[index_curr]
        counter = counter+1
    print ("Para um mínimo de n =",min_iter,"e epsilon =",eps)
    print (" n  | delta( T(n+1)/T(n) )")
    print (counter, " |", delta)
    print ("converge para ", curr)
    print ("")

In [31]:
# Testes
t(10**(-15))
t(10**(-16))

t_force(10**(-16), 44)
t_force(10**(-20), 100)


Para um epsilon = 1e-15
 n  | delta( T(n+1)/T(n) )
39  | 6.661338147750939e-16
converge para  1.8392867552141603

Para um epsilon = 1e-16
 n  | delta( T(n+1)/T(n) )
43  | 0.0
converge para  1.8392867552141612

Para um mínimo de n = 44 e epsilon = 1e-16
 n  | delta( T(n+1)/T(n) )
44  | 0.0
converge para  1.8392867552141612

Para um mínimo de n = 100 e epsilon = 1e-20
 n  | delta( T(n+1)/T(n) )
100  | 0.0
converge para  1.8392867552141612



---

### Resposta

**Converge para** 1.8392867552141612 **com a precisão** de 15 dígitos decimais (o 16º pode sofrer de problemas de arredondamento).

**Explicação do código e raciocínio:**

Primeiramente, implemento uma função `t(esp)` que vai gerando e armazenando os últimos 3 resultados da sequência de tribonacci de forma sequencial (o que tem performance melhor do que a forma recursiva).

Essa função calcula 

$delta = delta(n) = |f(n+1)-f(n)|$

para 

$f(n) = T_{n+1}/T_n$ 

a cada iteração e compara $delta$ com um valor limite $eps$. Quando $delta < eps$, o algoritmo convergiu.

Porém, na sequência de valores gerada por $delta(n)$, alguns valores eram $delta = 0$ seguidos de valores diferentes de $0$, o que levava o algoritmo a cnsiderar a convergência nas primeiras iterações. 

Portanto, implementei uma verificação para apenas parar o algoritmo no $delta = 0$ se o valor se repetir mais de uma vez seguida.

Implementei também `t_force(eps, min_iter)` que define com o parâmetro `min_iter` um valor mínimo para $n$, a fim de comprovar que, para a precisão do Python puro, o melhor valor que podemos obter é o obtido com com função `t(eps)`.

-----

Observe que calcular `tribonacci(30)` pode demorar um pouco. Isso porque a função recursiva se chama muitas vezes.

Seja $C_n$ o número de vezes que a função `tribonacci` é chamada para calcular $T_n$.
Escreva a recorrência que $C_n$ satisfaz (e as condições iniciais!).

Da maneira que foi implementada: 

$$ C_n = \cases {1 & se $n \in \{0,1,2\}$\\ 1 + C_{n-1} + C_{n-2} + C_{n-3} & se $n > 2$} $$

Agora, programe a função que *calcula* $C_n$!

In [22]:
def c(n):
    if (n in [0,1,2]): return 1
    elif (n>2): return 1 + c(n-1) + c(n-2) + c(n-3)

print(c(3))
print(c(4))
print(c(5))

4
7
13
